Fixes #71: actions buttons on job page extracted into their own component

m0mosenpai opened the pull request:
Closes #71 

- [X] Create a new reusable component
- [X] Add template code from existing job-result component
- [x] Implement https://akveo.github.io/nebular/docs/components/actions/overview#nbactionscomponent
- [x] Use new reusable component in Result page
- [x] Finalize and test everything

m0mosenpai commented:
![image](https://user-images.githubusercontent.com/17424778/112700471-3dfa3f80-8eb4-11eb-9073-baf8798a138d.png)
As I understand, we want this part to be made into a new separate component and then use it on the Results page and the Dashboard both. Where in the dashboard is it going to be displayed? In a separate column besides each completed job? Or somewhere else?

I have made this draft PR so I can approach this step by step and get some help as well since I'm new to Angular. Hope that's not an issue.

Eshaan7 commented:
> ![image](https://user-images.githubusercontent.com/17424778/112700471-3dfa3f80-8eb4-11eb-9073-baf8798a138d.png)
> As I understand, we want this part to be made into a new separate component and then use it on the Results page and the Dashboard both. Where in the dashboard is it going to be displayed?

Yes, those buttons are to be extracted into their own component. For now, I'd say let's not think about where to place it in the dashboard. The main aim is to break down the large JobResult component into seperate components so it's easier to maintain. 

So all you have to do is create that new component and use it inside JobResult. We can add it to the dashboard later.

m0mosenpai commented:
@Eshaan7 I moved the JobActions specific things in the HTML and CSS to the new component. Also moved the async functions in the .ts file over to the .ts file of the new component. I'm bit unsure now as to how the program flow will be and how I should connect the 2 components so that I can use this new one inside the old JobResults one.

Eshaan7 commented on src/app/@theme/components/job-actions-menu/job-actions-menu.component.ts:
```diff
@@ -0,0 +1,109 @@
+import { Job } from '../../../@core/models/models';
+
+@Component({
+  selector: 'intelowl-job-actions-menu',
```
`job-actions-menu` is enough for the name.

Eshaan7 commented on src/app/@theme/components/job-actions-menu/job-actions-menu.component.ts:
```diff
@@ -0,0 +1,109 @@
+  public jobId: number;
+
+  // interval var
+  private pollInterval: any;
```
all the code related to polling is not needed in this component.

Line 18, 37, 105

m0mosenpai replied:
what about the subscription code in `ngOnInit(): void` ? Should only the polling part be removed or the entire thing? And if we aren't subscribing, what will come in `ngOninit()` instead?

Eshaan7 replied:
the polling part will stay as is in JobResult and is not required in actions component. ngOnInit and ngOnDestroy can be removed from the actions component.

Eshaan7 commented on src/app/@theme/theme.module.ts:
```diff
@@ -28,6 +28,7 @@ import { DEFAULT_THEME } from './styles/theme.default';
 import { DARK_THEME } from './styles/theme.dark';
 
 import { FormsModule } from '@angular/forms';
+import { JobActionsMenuComponent } from './components/job-actions-menu/job-actions-menu.component';
```
The `job-actions-menu` folder/component should be placed inside of `/src/app/pages/dashboard/job-result/` folder i.e. the dashboard module, not the themes module.

Eshaan7 commented on src/app/pages/dashboard/job-result/job-result.component.html:
```diff
@@ -18,49 +18,7 @@ <h5 class="d-flex justify-content-between">
-                >
-                </nb-icon>
-              </span>
+              <intelowl-job-actions-menu> </intelowl-job-actions-menu>
```
need to pass the `job` object from JobResult to new component. Use @Input decorator on the new component for receiving this.

Eshaan7 replied:
> I'm bit unsure now as to how the program flow will be and how I should connect the 2 components so that I can use this new one inside the old JobResults one.

@m0mosenpai. You've used the HTML tag correctly. Just need to pass the job object from JobResult component to the actions component (using the angular's @Input decorator to receive the prop). Angular has bit of a learning curve, so make sure to read about the basics. Like, how to structure the DOM, how to pass data from components, @Input, @Output, @Event decorators, etc.

Eshaan7 commented:
Please make the requested changes and squash all commits into one, thank you.

m0mosenpai commented:
After making the above changes (minus the polling part), these were the results:
![image](https://user-images.githubusercontent.com/17424778/113148812-eceda100-924f-11eb-8882-f12279727926.png)

After job finished successfully,
![image](https://user-images.githubusercontent.com/17424778/113148922-127aaa80-9250-11eb-8e2f-1e626a56bb79.png)

I noticed that the green tick that's present to the left of these buttons after the job finished succesfully, is missing. It was there in the demo firebase app.
![image](https://user-images.githubusercontent.com/17424778/113149352-81f09a00-9250-11eb-8a44-de3b1869b3d2.png)

Eshaan7 commented on src/app/pages/dashboard/job-result/job-actions-menu/job-actions-menu.component.ts:
```diff
@@ -0,0 +1,108 @@
+  styleUrls: ['./job-actions-menu.component.scss'],
+})
+export class JobActionsMenuComponent implements OnInit, OnDestroy {
+  @Input() jobTableData: Job;
```
can u rename this to `@Input() job: Job` just for simplicity? Thanks.

Eshaan7 replied:
don't forget to make corresponding changes in HTML file.

Eshaan7 commented on src/app/pages/dashboard/job-result/job-result.component.html:
```diff
@@ -18,49 +18,9 @@ <h5 class="d-flex justify-content-between">
               </span>
-              <span>
-                <!-- Job Status Icon -->
-                <intelowl-job-status-icon
```
You should not have removed line 22 to line 28.

m0mosenpai replied:
Oh right, I must have missed this while moving the html code to the new component.

m0mosenpai commented:
During analysis:
![image](https://user-images.githubusercontent.com/17424778/113163892-38a74700-925e-11eb-9b89-b272638197bf.png)

After finishing:
![image](https://user-images.githubusercontent.com/17424778/113163955-478df980-925e-11eb-8e7d-bc0b04bc7b04.png)

Eshaan7 commented on src/app/pages/dashboard/job-result/job-actions-menu/job-actions-menu.component.ts:
```diff
@@ -0,0 +1,84 @@
+export class JobActionsMenuComponent {
+  @Input() job: Job;
+
+  // Job ID whose result is being displayed
+  public jobId: number;
```
not needed

Eshaan7 replied:
not needed

Eshaan7 replied:
lines 16, 17, 18, 19, 22, 23 are not needed.

Eshaan7 commented on src/app/pages/dashboard/job-result/job-actions-menu/job-actions-menu.component.ts:
```diff
@@ -0,0 +1,84 @@
+export class JobActionsMenuComponent {
+  @Input() job: Job;
+
+  // Job ID whose result is being displayed
+  public jobId: number;
+  // interval var
+  private pollInterval: any;
+  // RxJS Subscription
+  private sub: Subscription;
+  // if true, shows error template
+  public isError: boolean = false;
```
Lines 16-23 can be removed.

m0mosenpai replied:
The button functions use `jobId`. Will that be passed from the parent component as a prop then?

Eshaan7 replied:
Use `job.id`

Eshaan7 commented on src/app/pages/dashboard/job-result/job-actions-menu/job-actions-menu.component.ts:
```diff
@@ -0,0 +1,84 @@
+    private readonly toastr: ToastService,
+    private readonly router: Router
+  ) {
+    this.sub = this.activateRoute.params.subscribe(
+      (res) => (this.jobId = res.jobId)
+    );
```
not needed.

m0mosenpai commented on src/app/pages/dashboard/job-result/job-actions-menu/job-actions-menu.component.ts:
```diff
@@ -0,0 +1,70 @@
+  ) {}
+
+  async getJobSample(): Promise<void> {
+    const url: string = await this.jobService.downloadJobSample(this.jobId);
```
yeah sorry I realised that later ...had made the change already ..fixed it now

Eshaan7 commented:
Thank you for the contribution!!
